lib/CGameState.cpp

```diff
@@ -3349,68 +3346,73 @@ void CPathfinder::calculatePaths()
 			turn++;
 		}
 
-
 		//add accessible neighbouring nodes to the queue
 		neighbours.clear();
 
-		//handling subterranean gate => it's exit is the only neighbour
-        bool subterraneanEntry = (ct->topVisitableId() == Obj::SUBTERRANEAN_GATE && useSubterraneanGates);
-		if(subterraneanEntry)
+		auto sObj = ct->topVisitableObj(cp->coord == CGHeroInstance::convertPosition(hero->pos, false));
+		auto cObj = dynamic_cast<const CGTeleport *>(sObj);
+		if(gs->isTeleportEntrancePassable(cObj, hero->tempOwner)
+			&& (addTeleportWhirlpool(dynamic_cast<const CGWhirlpool *>(cObj))
+				|| addTeleportTwoWay(cObj)
+				|| addTeleportOneWay(cObj)
+				|| addTeleportOneWayRandom(cObj)))
 		{
-			//try finding the exit gate
-			if(const CGObjectInstance *outGate = getObj(CGTeleport::getMatchingGate(ct->visitableObjects.back()->id), false))
+			for(auto objId : gs->getTeleportChannelExits(cObj->channel, ObjectInstanceID(), hero->tempOwner))
 			{
-				const int3 outPos = outGate->visitablePos();
-				//gs->getNeighbours(*getTile(outPos), outPos, neighbours, boost::logic::indeterminate, !cp->land);
-				neighbours.push_back(outPos);
+				auto obj = getObj(objId);
+				if(CGTeleport::isExitPassable(gs, hero, obj))
+					neighbours.push_back(obj->visitablePos());
 			}
-			else
+		}
+
+		std::vector<int3> neighbour_tiles;
+		gs->getNeighbours(*ct, cp->coord, neighbour_tiles, boost::logic::indeterminate, !cp->land);
+		if(sObj)
+		{
+			for(int3 neighbour_tile: neighbour_tiles)
 			{
-				//gate with no exit (blocked) -> do nothing with this node
-				continue;
+				if(canMoveBetween(neighbour_tile, sObj->visitablePos()))
+					neighbours.push_back(neighbour_tile);
 			}
 		}
-
-		gs->getNeighbours(*ct, cp->coord, neighbours, boost::logic::indeterminate, !cp->land);
+		else
+			vstd::concatenate(neighbours, neighbour_tiles);
 
 		for(auto & neighbour : neighbours)
 		{
 			const int3 &n = neighbour; //current neighbor
 			dp = getNode(n);
 			dt = &gs->map->getTile(n);
-            destTopVisObjID = dt->topVisitableId();
+			destTopVisObjID = dt->topVisitableId();
 
 			useEmbarkCost = 0; //0 - usual movement; 1 - embark; 2 - disembark
-
-			int turnAtNextTile = turn;
-
-
 			const bool destIsGuardian = sourceGuardPosition == n;
 
-			if(!goodForLandSeaTransition())
-				continue;
-
-			if(!canMoveBetween(cp->coord, dp->coord) || dp->accessible == CGPathNode::BLOCKED )
+			auto dObj = dynamic_cast<const CGTeleport*>(dt->topVisitableObj());
```

Comment:
Here should be check for successfull cast or at least comment about that all methods above can accept nullptr.